test/raft/unit/test_snapshot.c

```diff
@@ -359,26 +399,95 @@ TEST(snapshot_leader, timeouts, set_up, tear_down, 0, NULL) {
 
 	PRE(sm_state(&leader.sm) == LS_PAGE_READ);
 	ut_rpc_sent(&leader.rpc);
+	munit_assert_int(ut_get_msg_sent().type, ==, RAFT_IO_INSTALL_SNAPSHOT_CP);
 	ut_rpc_to_expired(&leader.rpc);
 
 	PRE(sm_state(&leader.sm) == LS_PAGE_READ);
 	ut_rpc_sent(&leader.rpc);
+	munit_assert_int(ut_get_msg_sent().type, ==, RAFT_IO_INSTALL_SNAPSHOT_CP);
 	ut_leader_message_received(&leader, ut_page_result());
 
 	PRE(sm_state(&leader.sm) == LS_SNAP_DONE);
 	ut_rpc_sent(&leader.rpc);
+	munit_assert_int(ut_get_msg_sent().type, ==, RAFT_IO_INSTALL_SNAPSHOT);
 	ut_leader_message_received(&leader, ut_install_snapshot_result());
 
 	sm_fini(&leader.sm);
 	return MUNIT_OK;
 }
 
+struct test_fixture {
+	union {
+		struct leader leader;
+		struct follower follower;
+	};
+	/* true when union contains leader, false when it contains follower */
+	bool is_leader;
+
+	/* We only expect one message to be in-flight. */
+	struct raft_message last_msg_sent;
+	bool msg_valid;
+
+	pool_t pool;
+
+	/* TODO: should accomodate several background jobs in the future. Probably
+	 * by scheduling a barrier in the pool after all the works that toggles this
+	 * flag. */
+	work_op orig_work_cb;
+	bool work_done;
+};
+
+/* Not problematic because each test runs in a different process. */
+static struct test_fixture global_fixture;
+
+#define MAGIC_MAIN_THREAD 0xdef1
+
+static __thread int thread_identifier;
+
+static void *pool_set_up(MUNIT_UNUSED const MunitParameter params[],
+                   MUNIT_UNUSED void *user_data)
+{
+	/* Prevent hangs. */
+	alarm(2);
+
+	pool_init(&global_fixture.pool, uv_default_loop(), 4, POOL_QOS_PRIO_FAIR);
+	global_fixture.pool.flags |= POOL_FOR_UT;
+	thread_identifier = MAGIC_MAIN_THREAD;
+
+	struct fixture *f = munit_malloc(sizeof *f);
+	return f;
+}
+
+static void pool_tear_down(void *data)
+{
+    free(data);
```

Comment:
`pool_close()` and `pool_fini()` are required.

Question: Do we use "ASAN" builds to test the protocol?

Reply:
Yes, we are compiling the file with `-fsanitize=address` but for some reason it is not picking up on leaked resources allocated in the tests. I will look into that.

Reply:
Okay, now I now why. The tests are in fact compiled with ASAN but `pool_init` does not reserve any memory so it is not noticing that.